### PR TITLE
Migration for shell workable tests to e2e

### PR DESCRIPTION
**Pull request by NeerajNagure**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR migratees shell workable tests to e2e
#### Which issue(s) this PR fixes:

xref #1062 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by NeerajNagure** [comment]
@wzshiming Please review this PR.I have also updated the testdata that will be changed due to adding some more arguments in the cluster.

File: test/e2e/workable.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+			if err = testPrometheus(); err != nil {
+				t.Fatal(err)
+			}
+			if err = testJaeger(); err != nil {
+				t.Fatal(err)
+			}
+			if err = testKwokControllerPort(); err != nil {
+				t.Fatal(err)
+			}
+			return ctx
```
Comment (wzshiming) [inline test/e2e/workable.go:317]: We should make sure that the test style is the same
- each step is wrapped in Assess()
- using the same retry mechanism as the other tests
- Use (*testing.T).Log instead fmt.Println

Comment (NeerajNagure) [reply test/e2e/workable.go:317]: @wzshiming  By each step do you mean each function i.e testWorkable,testKubeSchedulerPort,testPrometheus,testJaeger,etc. should get a separate assess step?

**Comment by NeerajNagure** [comment]
@wzshiming I have made the changes.Please take a look

File: test/e2e/workable.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	output, err = exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfigPath, "get", "pod").Output() // #nosec G204
+	if err != nil {
+		return err
+	}
+	if !strings.Contains(string(output), "Running") {
+		return fmt.Errorf("kubeconfig not working")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:79]:
```suggestion
func testWorkable(ctx context.Context, kwokctlPath, name string) error {
	cmd := exec.Command("kubectl", "config", "current-context")
	currentContext, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("error getting current context: %w", err)
	}
	if strings.TrimSpace(string(currentContext)) != "kwok-"+name {
		return fmt.Errorf("current context is %s, expected kwok-%s", currentContext, name)
	}
	err = exec.CommandContext(ctx, kwokctlPath, "--name", name, "scale", "node", "fake-node", "--replicas=1").Run()
	if err != nil {
		return fmt.Errorf("failed to scale node: %w", err)
	}
	err = exec.CommandContext(ctx, kwokctlPath, "--name", name, "scale", "pod", "fake-pod", "--replicas=1").Run()
	if err != nil {
		return fmt.Errorf("failed to scale pod: %w", err)
	}
	output, err := exec.CommandContext(ctx, kwokctlPath, "--name", name, "kubectl", "get", "pod").Output()
	if err != nil {
		return err
	}

	// TODO: Use json output instead and check that node and pod work as expected
	if !strings.Contains(string(output), "Running") {
		return fmt.Errorf("pod not running")
	}
	return nil
}

func testGetKubeconfig(ctx context.Context, kwokctlPath, name string) error {
	kubeconfigPath := name + ".kubeconfig"
	output, err := exec.CommandContext(ctx, kwokctlPath, "--name", name, "get", "kubeconfig", "--user", "cluster-admin", "--group", "system:masters").Output()
	if err != nil {
		return fmt.Errorf("failed to get kubeconfig: %w", err)
	}
	if err := os.WriteFile(kubeconfigPath, output, 0640); err != nil {
		return fmt.Errorf("failed to write kubeconfig: %w", err)
	}
	defer func() {
		_ = os.Remove(kubeconfigPath)
	}()

	output, err = exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfigPath, "get", "pod").Output() // #nosec G204
	if err != nil {
		return err
	}

	// TODO: Use json output instead and check that pod work as expected
	if !strings.Contains(string(output), "Running") {
		return fmt.Errorf("kubeconfig not working")
	}
	return nil
}
```

File: test/e2e/workable.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	if err != nil {
+		return err
+	}
+	targets = string(body)
+	if strings.Count(targets, `"health":"up"`) < 6 {
+		return fmt.Errorf("not enough healthy targets")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:104]:
```suggestion
func testPrometheus() error {
	resp, err := http.Get("http://127.0.0.1:9090/api/v1/targets")
	if err != nil {
		return fmt.Errorf("failed to get targets: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	// TODO: checks that the healthy component is consistent with the actual component being started
	if strings.Count(string(body), `"health":"up"`) < 6 {
		return fmt.Errorf("not enough healthy targets")
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+	if resp.StatusCode == http.StatusOK {
+		return nil
+	}
+	return fmt.Errorf("jaeger not healthy")
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:125]:
```suggestion
func testJaeger() error {
	resp, err := http.Get("http://127.0.0.1:16686/api/services")
	if err != nil {
		return err
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}

	// TODO: Check that the output is expected
	return nil
}
```

File: test/e2e/workable.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+	if string(body) != "ok" {
+		t.Log("Error: controller healthz is not ok")
+		t.Log("curl -s http://127.0.0.1:10247/healthz")
+		t.Log(string(body))
+		return fmt.Errorf("controller healthz is not ok")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:151]:
```suggestion
func testKwokControllerPort() error {
	resp, err := http.Get("http://127.0.0.1:10247/healthz")
	if err != nil {
		return fmt.Errorf("controller healthz check failed: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if string(body) != "ok" {
		return fmt.Errorf("output is not unexpected: %q", string(body))
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+	if !strings.Contains(string(body), `"health"`) {
+		t.Log("Error: etcd connection")
+		t.Log("curl -s http://127.0.0.1:2400/health")
+		t.Log(string(body))
+		return fmt.Errorf("etcd connection error")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:177]:
```suggestion
func testEtcdPort() error {
	resp, err := http.Get("http://127.0.0.1:2400/health")
	if err != nil {
		return fmt.Errorf("etcd health check failed: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if !strings.Contains(string(body), `"health"`) {
		return fmt.Errorf("output is not unexpected: %q", string(body))
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to get namespace(default) by kwokctl etcdctl in cluster %s: %w", name, err)
+	}
+
+	if !strings.Contains(string(output), "default") {
+		return fmt.Errorf("namespace(default) not found in cluster %s", name)
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:190]:
```suggestion
func testEtcdctlGet(ctx context.Context, kwokctlPath, name string) error {
	output, err := exec.CommandContext(ctx, kwokctlPath, "--name="+name, "etcdctl", "get", "/registry/namespaces/default", "--keys-only").Output() // #nosec G204
	if err != nil {
		return fmt.Errorf("failed to get namespace(default) by kwokctl etcdctl in cluster %s: %w", name, err)
	}

	if !strings.Contains(string(output), "default") {
		return fmt.Errorf("namespace(default) not found in cluster %s", name)
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+	if string(body) != "ok" {
+		t.Log("Error: kube scheduler connection")
+		t.Log("curl -s  https://127.0.0.1:10250/healthz")
+		t.Log(string(body))
+		return fmt.Errorf("kube scheduler connection error")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:216]:
```suggestion
func testKubeSchedulerPort() error {
	// TODO: get CA from kwok cluster
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // #nosec G402
		},
	}
	resp, err := client.Get("https://127.0.0.1:10250/healthz")
	if err != nil {
		return fmt.Errorf("kube scheduler connection error: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if string(body) != "ok" {
		return fmt.Errorf("output is not unexpected: %q", string(body))
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+	if string(body) != "ok" {
+		t.Log("Error: kube controller manager connection")
+		t.Log("curl -s https://127.0.0.1:10260/healthz")
+		t.Log(string(body))
+		return fmt.Errorf("kube controller manager connection error")
+	}
+
+	return nil
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:242]:
```suggestion
func testKubeControllerManagerPort() error {
	// TODO: get CA from kwok cluster
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // #nosec G402
		},
	}
	resp, err := client.Get("https://127.0.0.1:10260/healthz")
	if err != nil {
		return fmt.Errorf("kube controller manager connection error: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status is not 200 OK")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if string(body) != "ok" {
		return fmt.Errorf("output is not unexpected: %q", string(body))
	}

	return nil
}
```

File: test/e2e/workable.go [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+			defer cancel()
+			return ctx
+		}).
+		Assess("Test Kwok controller port", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+			if err := testKwokControllerPort(t); err != nil {
+				t.Fatal(err)
+			}
+			return ctx
+		})
+}
```
Comment (wzshiming) [inline test/e2e/workable.go:322]:
```suggestion
func CaseWorkable(kwokctlPath, clusterName, clusterRuntime string) *features.FeatureBuilder {
	f0 := features.New("Workable").
		Assess("Test Workable", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
			err := testWorkable(ctx, kwokctlPath, clusterName)
			if err != nil {
				t.Fatal(err)
			}
			return ctx
		}).
		Assess("Test Get kubeconfig", func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
			if err := testGetKubeconfig(ctx, kwokctlPath, clusterName); err != nil {
				t.Fatal(err)
			}
			return ctx
		})

	if clusterRuntime != "kind" && clusterRuntime != "kind-podman" {
		f0 = f0.
			Assess("Test Kube Controller Manager Port", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
				if err := testKubeControllerManagerPort(); err != nil {
					t.Fatal(err)
				}
				return ctx
			}).
			Assess("Test Kube Scheduler Port", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
				if err := testKubeSchedulerPort(); err != nil {
					t.Fatal(err)
				}
				return ctx
			}).
			Assess("Test Kube Etcdctl Port", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
				if err := testEtcdPort(); err != nil {
					t.Fatal(err)
				}
				return ctx
			})
	}

	if runtime.GOOS != "windows" {
		f0 = f0.Assess("Test Etcdctl Get", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
			if err := testEtcdctlGet(ctx, kwokctlPath, clusterName); err != nil {
				t.Fatal(err)
			}
			return ctx
		})
	}

	f0 = f0.
		Assess("Test Prometheus", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
			if err := testPrometheus(); err != nil {
				t.Fatal(err)
			}
			return ctx
		}).
		Assess("Test Jaeger", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
			if err := testJaeger(); err != nil {
				t.Fatal(err)
			}
			return ctx
		}).
		Assess("Test Kwok controller port", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
			if err := testKwokControllerPort(); err != nil {
				t.Fatal(err)
			}
			return ctx
		})

	// TODO: Check that the other components work
	return f0
}
```

**Review by wzshiming — COMMENTED** [review]
Some style adjustments
And I've left a few TODO items, better to be done, optional.

**Comment by NeerajNagure** [comment]
@wzshiming I think we need to set the context first because the current context is not set otherwise the tests will fail

File: test/e2e/workable.go [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+	"runtime"
+	"strings"
+	"testing"
+
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+)
+
+func testWorkable(ctx context.Context, kwokctlPath, name string) error {
+	cmd := exec.Command("kubectl", "config", "current-context")
```
Comment (wzshiming) [inline test/e2e/workable.go:36]: > workable.go:236: exec: "kubectl": executable file not found in $PATH

https://github.com/kubernetes-sigs/kwok/actions/runs/10561061180/job/29255960861?pr=1195


Add `${rootDir}/bin` to $PATH in `main_test.go`, I remember kubectl will download there.

**Comment by NeerajNagure** [comment]
@wzshiming The issue has been fixed.

**Comment by wzshiming** [comment]
LGTM, then squash all the commits

**Comment by NeerajNagure** [comment]
@wzshiming I have squashed the commits

**Comment by wzshiming** [comment]
The pull-kwok-e2e-test-main was always stuck in TestBenchmarking/Benchmarking/Create_pods

**Pull request merged**